`_start_opensearch` does not catch

There seems to be some race condition happening at `_post_start_init`, where some of the CI tests fail with `OpenSearchUserMgmtError`. That is caused in the stack at the end.

Side note: `_post_start_init` should be renamed to something else, as it runs as the very first line at `_start_opensearch`.

Stack trace: https://pastebin.ubuntu.com/p/WRMY3FnZCc/
```
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_distro.py", line 276, in call
    response = s.request(**request_kwargs)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/requests/adapters.py", line 532, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='10.206.26.175', port=9200): Read timed out. (read timeout=5)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_users.py", line 137, in get_users
    return self.opensearch.request("GET", f"{USER_ENDPOINT}/")
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_distro.py", line 297, in request
    resp = call(retries, resp_status_code)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_distro.py", line 285, in call
    return call(
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_distro.py", line 238, in call
    raise OpenSearchHttpError()
charms.opensearch.v0.opensearch_exceptions.OpenSearchHttpError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-opensearch-4/charm/./src/charm.py", line 94, in <module>
    main(OpenSearchOperatorCharm)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/ops/main.py", line 434, in main
    framework.reemit()
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/ops/framework.py", line 863, in reemit
    self._reemit()
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/ops/framework.py", line 942, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/rolling_ops/v0/rollingops.py", line 410, in _on_run_with_lock
    callback(event)
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 618, in _start_opensearch
    self._post_start_init()
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 720, in _post_start_init
    self._put_monitoring_user()
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_base_charm.py", line 840, in _put_monitoring_user
    users = self.user_manager.get_users()
  File "/var/lib/juju/agents/unit-opensearch-4/charm/lib/charms/opensearch/v0/opensearch_users.py", line 139, in get_users
    raise OpenSearchUserMgmtError(e)
charms.opensearch.v0.opensearch_users.OpenSearchUserMgmtError
2024-01-24 08:12:51 ERROR juju.worker.uniter.operation runhook.go:180 hook "update-status" (via hook dispatching script: dispatch) failed: exit status 1
```

So, I did some more digging on this issue.

In a nutshell, I believe we try to access the .opensearch-security index, which is hosted on unit opensearch/3, at the moment that unit is restarting.

The main question here is where was the valid replicas of `.opensearch-security` at the moment `_post_create_monitoring_user()` is called in `_post_start_init`.

---------------------------------------------------------------------------------------------------------------------------------------------------

There are 3x OpenSearch nodes:
```
Model              Controller           Cloud/Region         Version  SLA          Timestamp
test-backups-xhv5  localhost-localhost  localhost/localhost  3.1.7    unsupported  11:38:45+01:00

App                       Version  Status  Scale  Charm                     Channel      Rev  Exposed  Message
opensearch                         active      3  opensearch                               1  no       
s3-integrator                      active      1  s3-integrator             latest/edge   13  no       
self-signed-certificates           active      1  self-signed-certificates  latest/edge   60  no       

Unit                         Workload  Agent  Machine  Public address  Ports  Message
opensearch/3*                active    idle   5        10.206.26.226          
opensearch/4                 active    idle   6        10.206.26.175          
opensearch/5                 active    idle   7        10.206.26.254          
s3-integrator/0*             active    idle   0        10.206.26.246          
self-signed-certificates/0*  active    idle   4        10.206.26.139          

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.206.26.246  juju-c72925-0  ubuntu@22.04      Running
4        started  10.206.26.139  juju-c72925-4  ubuntu@22.04      Running
5        started  10.206.26.226  juju-c72925-5  ubuntu@22.04      Running
6        started  10.206.26.175  juju-c72925-6  ubuntu@22.04      Running
7        started  10.206.26.254  juju-c72925-7  ubuntu@22.04      Running
```

The node having an issue is `opensearch/4`, which starts to have issues accessing the `internalusers` API on `08:12:15`:
```
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 https://10.206.26.175:9200 "GET /_cluster/settings HTTP/1.1" 200 182
2024-01-24 08:12:15 ERROR unit.opensearch/4.juju-log server.go:325 [Errno 111] Connection refused
2024-01-24 08:12:15 ERROR unit.opensearch/4.juju-log server.go:325 [Errno 111] Connection refused
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Getting secret app:admin-password
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 https://10.206.26.175:9200 "PUT /_cluster/settings HTTP/1.1" 200 125
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Getting secret app:admin-password
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 https://10.206.26.175:9200 "PUT /.plugins-ml-config/_settings HTTP/1.1" 200 21
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Getting secret app:admin-password
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 https://10.206.26.175:9200 "PUT /.opensearch-sap-log-types-config/_settings HTTP/1.1" 200 21
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Getting secret app:admin-password
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 https://10.206.26.175:9200 "PUT /.opensearch-sap-pre-packaged-rules-config/_settings HTTP/1.1" 404 575
2024-01-24 08:12:15 ERROR unit.opensearch/4.juju-log server.go:325 [Errno 111] Connection refused
2024-01-24 08:12:15 ERROR unit.opensearch/4.juju-log server.go:325 [Errno 111] Connection refused
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Getting secret app:admin-password
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 https://10.206.26.175:9200 "GET /_cluster/health HTTP/1.1" 200 447
2024-01-24 08:12:15 INFO unit.opensearch/4.juju-log server.go:325 Health: {'cluster_name': 'opensearch-5raf', 'status': 'red', 'timed_out': False, 'number_of_nodes': 1, 'number_of_data_nodes': 1, 'discovered_master': True, 'discovered_cluster_manager': True, 'activ>
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Getting secret app:admin-password
2024-01-24 08:12:15 DEBUG unit.opensearch/4.juju-log server.go:325 Starting new HTTPS connection (1): 10.206.26.175:9200
2024-01-24 08:12:20 ERROR unit.opensearch/4.juju-log server.go:325 Request GET to https://10.206.26.175:9200/_plugins/_security/api/internalusers/ with payload: None failed. (Attempts left: 0)
HTTPSConnectionPool(host='10.206.26.175', port=9200): Read timed out. (read timeout=5)
2024-01-24 08:12:21 ERROR unit.opensearch/4.juju-log server.go:325 Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-opensearch-4/charm/venv/urllib3/connectionpool.py", line 467, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from

###### STARTS THE STACK TRACE MENTIONED IN THIS BUG
```
On the logs above, it is possible to see that multiple calls actually go through, but eventually the `GET https://10.206.26.175:9200/_plugins/_security/api/internalusers/` fails. Also, the health status is marked as "RED".

Looking at the logs of opensearch/4, I can see the unit loses contact with `opensearch/3` and eventually decides to remove it from the cluster:
```
[2024-01-24T08:11:37,496][WARN ][o.o.c.NodeConnectionsService] [opensearch-4] failed to connect to {opensearch-3}{nWI1aoofTkasg0WYc3Wc1A}{mytOU1xoSYWg5xJOS-CAlg}{10.206.26.226}{10.206.26.226:9300}{coordinating_onlydimml}{shard_indexing_pressure_enabled=true} 
org.opensearch.transport.ConnectTransportException: [opensearch-3][10.206.26.226:9300] connect_exception
        at org.opensearch.transport.TcpTransport$ChannelsConnectedListener.onFailure(TcpTransport.java:1074) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.core.action.ActionListener.lambda$toBiConsumer$2(ActionListener.java:217) ~[opensearch-core-2.10.0.jar:2.10.0]
        at org.opensearch.common.concurrent.CompletableContext.lambda$addListener$0(CompletableContext.java:57) ~[opensearch-common-2.10.0.jar:2.10.0]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2325) ~[?:?]
        at org.opensearch.common.concurrent.CompletableContext.addListener(CompletableContext.java:60) ~[opensearch-common-2.10.0.jar:2.10.0]
        at org.opensearch.transport.netty4.Netty4TcpChannel.addConnectListener(Netty4TcpChannel.java:135) ~[?:?]
        at org.opensearch.transport.TcpTransport.initiateConnection(TcpTransport.java:378) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.transport.TcpTransport.openConnection(TcpTransport.java:338) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.transport.ClusterConnectionManager.internalOpenConnection(ClusterConnectionManager.java:274) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.transport.ClusterConnectionManager.connectToNode(ClusterConnectionManager.java:157) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.transport.TransportService.connectToNode(TransportService.java:493) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.transport.TransportService.connectToNode(TransportService.java:473) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.cluster.NodeConnectionsService$ConnectionTarget$1.doRun(NodeConnectionsService.java:344) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-2.10.0.jar:2.10.0]
        at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
        at org.opensearch.cluster.NodeConnectionsService.connectToNodes(NodeConnectionsService.java:157) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.cluster.service.ClusterApplierService.connectToNodesAndWait(ClusterApplierService.java:583) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:548) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:484) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:186) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:849) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedOpenSearchThreadPoolExecutor.java:282) ~[opensearch-2.10.0.jar:2.10.0]
        at org.opensearch.common.util.concurrent.PrioritizedOpenSearchThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedOpenSearchThreadPoolExecutor.java:245) ~[opensearch-2.10.0.jar:2.10.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: 10.206.26.226/10.206.26.226:9300
Caused by: java.net.ConnectException: Connection refused
        at sun.nio.ch.Net.pollConnect(Native Method) ~[?:?]
        at sun.nio.ch.Net.pollConnectNow(Net.java:672) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946) ~[?:?]
        at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:337) ~[?:?]
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:689) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:652) ~[?:?]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[?:?]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[?:?]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[?:?]
        ... 1 more
```

Then, the unit is removed:
```
[2024-01-24T08:11:37,582][INFO ][o.o.c.s.ClusterApplierService] [opensearch-4] removed {{opensearch-3}{nWI1aoofTkasg0WYc3Wc1A}{mytOU1xoSYWg5xJOS-CAlg}{10.206.26.226}{10.206.26.226:9300 {coordinating_onlydimml{shard_indexing_pressure_enabled=true}}, term: 2, version: 44, reason: Publication{term=2, version=44}                                                                                        
```

Now, looking into unit `opensearch/3`, indeed the unit is restarted at that time:
```
[2024-01-24T08:11:37,037][INFO ][o.o.d.PeerFinder         ] [opensearch-3] setting findPeersInterval to [1s] as node commission stat
us = [true] for local node [{opensearch-3}{nWI1aoofTkasg0WYc3Wc1A}{mytOU1xoSYWg5xJOS-CAlg}{10.206.26.226}{10.206.26.226:9300}{coordi
nating_onlydimml}{shard_indexing_pressure_enabled=true}]
[2024-01-24T08:11:37,131][INFO ][o.o.a.u.d.DestinationMigrationCoordinator] [opensearch-3] Detected cluster change event for destina
tion migration
[2024-01-24T08:11:37,212][INFO ][o.o.a.u.d.DestinationMigrationCoordinator] [opensearch-3] Detected cluster change event for destina
tion migration
[2024-01-24T08:11:37,345][INFO ][o.o.i.r.RecoverySourceHandler] [opensearch-3] [.opensearch-observability][0][recover to opensearch-
4] finalizing recovery took [44.6ms]
[2024-01-24T08:11:37,351][INFO ][o.o.i.r.RecoverySourceHandler] [opensearch-3] [.plugins-ml-config][0][recover to opensearch-4] fina
lizing recovery took [22.7ms]
Stopping Service for snap application opensearch.daemon...
[2024-01-24T08:11:37,398][INFO ][o.o.a.u.d.DestinationMigrationCoordinator] [opensearch-3] Detected cluster change event for destina
tion migration
[2024-01-24T08:11:37,405][INFO ][o.o.n.Node               ] [opensearch-3] stopping ...
[2024-01-24T08:11:37,421][INFO ][o.o.s.a.r.AuditMessageRouter] [opensearch-3] Closing AuditMessageRouter
[2024-01-24T08:11:37,422][INFO ][o.o.s.a.s.SinkProvider   ] [opensearch-3] Closing DebugSink
[2024-01-24T08:11:37,425][INFO ][o.o.c.c.Coordinator      ] [opensearch-3] cluster-manager node [{opensearch-4}{CxJCO-5xTRiPaEnbQk2p
2w}{ELLVDYRMSAqC9rpXIzr0Qg}{10.206.26.175}{10.206.26.175:9300}{coordinating_onlydimml}{shard_indexing_pressure_enabled=true}] failed
, restarting discovery
org.opensearch.transport.NodeDisconnectedException: [opensearch-4][10.206.26.175:9300][disconnected] disconnected
[2024-01-24T08:11:37,500][INFO ][o.o.n.Node               ] [opensearch-3] stopped
[2024-01-24T08:11:37,501][INFO ][o.o.n.Node               ] [opensearch-3] closing ...
[2024-01-24T08:11:37,515][INFO ][o.o.s.a.i.AuditLogImpl   ] [opensearch-3] Closing AuditLogImpl
[2024-01-24T08:11:37,527][INFO ][o.o.n.Node               ] [opensearch-3] closed
```

and only comes back on 08:12:30:
```
[2024-01-24T08:12:30,154][INFO ][o.o.n.Node               ] [opensearch-3] started
[2024-01-24T08:12:30,155][INFO ][o.o.s.OpenSearchSecurityPlugin] [opensearch-3] Node started
```

So, `PUT /<index-name>` commands will work, as it means creating new indices, but GET on a node that does not hold a replica of a given index will fail.

--- a/lib/charms/opensearch/v0/helper_cluster.py
+++ b/lib/charms/opensearch/v0/helper_cluster.py
@@ -58,6 +58,23 @@ class ClusterTopology:
         return base_roles + ["cluster_manager"]
 
     @staticmethod
+    def get_cluster_settings(
+        opensearch: OpenSearchDistribution,
+        host: Optional[str] = None,
+        alt_hosts: Optional[List[str]] = None,
+        include_defaults: bool = False,
+    ) -> Dict[str, any]:
+        """Get the cluster settings."""
+        settings = opensearch.request(
+            "GET",
+            f"/_cluster/settings?flat_settings=true&include_defaults={str(include_defaults).lower()}",
+            host=host,
+            alt_hosts=alt_hosts,
+        )
+
+        return dict(settings["defaults"] | settings["persistent"] | settings["transient"])
+
+    @staticmethod
     def recompute_nodes_conf(app_name: str, nodes: List[Node]) -> Dict[str, Node]:
         """Recompute the configuration of all the nodes (cluster set to auto-generate roles)."""
         # in case the cluster nodes' roles were previously "manually generated" - we need

--- a/lib/charms/opensearch/v0/opensearch_backups.py
+++ b/lib/charms/opensearch/v0/opensearch_backups.py
@@ -3,8 +3,8 @@
 
 """OpenSearch Backup.
 
-This file holds the implementation of the OpenSearchBackup, OpenSearchBackupPlugin classes
-as well as the configuration and state enum.
+This file holds the implementation of the OpenSearchBackup class, as well as the state enum
+and configuration.
 
 The OpenSearchBackup class listens to both relation changes from S3_RELATION and API calls
 and responses. The OpenSearchBackupPlugin holds the configuration info. The classes together
@@ -467,7 +467,7 @@ class OpenSearchBackup(Object):
             return BackupServiceState.RESPONSE_FAILED_NETWORK
         return self.get_service_status(output)
 
-    def _on_s3_credentials_changed(self, event: EventBase) -> None:
+    def _on_s3_credentials_changed(self, event: EventBase) -> None:  # noqa: C901
         """Calls the plugin manager config handler.
 
         This method will iterate over the s3 relation and check:
@@ -489,8 +489,14 @@ class OpenSearchBackup(Object):
         self.charm.status.set(MaintenanceStatus(BackupSetupStart))
 
         try:
+            if not self.charm.plugin_manager.check_plugin_manager_ready():
+                raise OpenSearchNotFullyReadyError()
+
             plugin = self.charm.plugin_manager.get_plugin(OpenSearchBackupPlugin)
             if self.charm.plugin_manager.status(plugin) == PluginState.ENABLED:
+                # We need to explicitly disable the plugin before reconfiguration
+                # That happens because, differently from the actual configs, we cannot
+                # retrieve the key values and check if they changed.
                 self.charm.plugin_manager.apply_config(plugin.disable())
             self.charm.plugin_manager.apply_config(plugin.config())
         except OpenSearchError as e:
@@ -507,6 +513,7 @@ class OpenSearchBackup(Object):
             PluginState.ENABLED,
             PluginState.WAITING_FOR_UPGRADE,
         ]:
+            logger.warning("_on_s3_credentials_changed: plugin is not enabled.")
             event.defer()
             return
 

--- a/lib/charms/opensearch/v0/opensearch_base_charm.py
+++ b/lib/charms/opensearch/v0/opensearch_base_charm.py
@@ -493,22 +493,21 @@ class OpenSearchBaseCharm(CharmBase):
             event.defer()
             return
 
-        self.status.set(MaintenanceStatus(PluginConfigStart))
         try:
+            self.status.set(MaintenanceStatus(PluginConfigStart))
             if self.plugin_manager.run():
                 self.on[self.service_manager.name].acquire_lock.emit(
                     callback_override="_restart_opensearch"
                 )
-        except OpenSearchNotFullyReadyError:
-            logger.warning("Plugin management: cluster not ready yet at config changed")
-            event.defer()
-            return
-        except OpenSearchPluginError:
-            self.status.set(BlockedStatus(PluginConfigChangeError))
+            self.status.clear(PluginConfigStart)
+        except (OpenSearchNotFullyReadyError, OpenSearchPluginError) as e:
+            if isinstance(e, OpenSearchNotFullyReadyError):
+                logger.warning("Plugin management: cluster not ready yet at config changed")
+            else:
+                self.status.set(BlockedStatus(PluginConfigChangeError))
             event.defer()
             return
         self.status.clear(PluginConfigChangeError)
-        self.status.clear(PluginConfigStart)
 
     def _on_set_password_action(self, event: ActionEvent):
         """Set new admin password from user input or generate if not passed."""
@@ -619,7 +618,6 @@ class OpenSearchBaseCharm(CharmBase):
                 event.defer()
                 self.defer_trigger_event.emit()
             return
-
         if not self._can_service_start():
             self.peers_data.delete(Scope.UNIT, "starting")
             event.defer()
@@ -632,9 +630,7 @@ class OpenSearchBaseCharm(CharmBase):
             self.peers_data.delete(Scope.UNIT, "starting")
             event.defer()
             return
-
         self.unit.status = WaitingStatus(WaitingToStart)
-
         rel = self.model.get_relation(PeerRelationName)
         for unit in rel.units.union({self.unit}):
             if rel.data[unit].get("starting") == "True":
@@ -646,7 +642,6 @@ class OpenSearchBaseCharm(CharmBase):
         try:
             # Retrieve the nodes of the cluster, needed to configure this node
             nodes = self._get_nodes(False)
-
             # validate the roles prior to starting
             self.opensearch_peer_cm.validate_roles(nodes, on_new_unit=True)
 

--- a/lib/charms/opensearch/v0/opensearch_plugin_manager.py
+++ b/lib/charms/opensearch/v0/opensearch_plugin_manager.py
@@ -10,9 +10,12 @@ This class is instantiated at the operator level and is called at every relevant
 config-changed, upgrade, s3-credentials-changed, etc.
 """
 
+import copy
+import functools
 import logging
-from typing import Any, Dict, List, Optional
+from typing import Any, Dict, List, Optional, Tuple
 
+from charms.opensearch.v0.helper_cluster import ClusterTopology
 from charms.opensearch.v0.opensearch_exceptions import (
     OpenSearchCmdError,
     OpenSearchNotFullyReadyError,
@@ -78,6 +81,11 @@ class OpenSearchPluginManager:
         self._keystore = OpenSearchKeystore(self._charm)
         self._event_scope = OpenSearchPluginEventScope.DEFAULT
 
+    @functools.cached_property
+    def cluster_config(self):
+        """Returns the cluster configuration."""
+        return ClusterTopology.get_cluster_settings(self._charm.opensearch, include_defaults=True)
+
     def set_event_scope(self, event_scope: OpenSearchPluginEventScope) -> None:
         """Sets the event scope of the plugin manager.
 
@@ -129,25 +137,31 @@ class OpenSearchPluginManager:
             }
         return {**self._charm_config, "opensearch-version": self._opensearch.version}
 
+    def check_plugin_manager_ready(self) -> bool:
+        """Checks if the plugin manager is ready to run."""
+        return (
+            self._charm.opensearch.is_node_up()
+            and len(self._charm._get_nodes(True)) == self._charm.app.planned_units()
+            and self._charm.health.apply()
+            in [
+                HealthColors.GREEN,
+                HealthColors.YELLOW,
+            ]
+        )
+
     def run(self) -> bool:
         """Runs a check on each plugin: install, execute config changes or remove.
 
         This method should be called at config-changed event. Returns if needed restart.
         """
-        if not self._charm.opensearch.is_started() or self._charm.health.apply() not in [
-            HealthColors.GREEN,
-            HealthColors.YELLOW,
-        ]:
-            # If the health is not green, then raise a cluster-not-ready error
-            # The classes above should then defer their own events in waiting.
-            # Defer is important as next steps to configure plugins will involve
-            # calls to the APIs of the cluster.
-            logger.info("Cluster not ready, wait for the next event...")
+        if not self.check_plugin_manager_ready():
             raise OpenSearchNotFullyReadyError()
 
         err_msgs = []
         restart_needed = False
         for plugin in self.plugins:
+            logger.info(f"Checking plugin {plugin.name}...")
+            logger.debug(f"Status: {self.status(plugin)}")
             # These are independent plugins.
             # Any plugin that errors, if that is an OpenSearchPluginError, then
             # it is an "expected" error, such as missing additional config; should
@@ -172,8 +186,8 @@ class OpenSearchPluginManager:
                 # This is a more serious issue, as we are missing some input from
                 # the user. The charm should block.
                 err_msgs.append(str(e))
-            except OpenSearchPluginError as e:
-                logger.error(f"Plugin {plugin.name} failed: {str(e)}")
+            logger.debug(f"Finished Plugin {plugin.name} status: {self.status(plugin)}")
+        logger.info(f"Plugin check finished, restart needed: {restart_needed}")
 
         if err_msgs:
             raise OpenSearchPluginError("\n".join(err_msgs))
@@ -255,6 +269,39 @@ class OpenSearchPluginManager:
         except KeyError as e:
             raise OpenSearchPluginMissingConfigError(plugin.name, configs=[f"{e}"])
 
+    def _compute_settings(
+        self, config: OpenSearchPluginConfig
+    ) -> Tuple[Dict[str, str], Dict[str, str]]:
+        """Returns the current and the new configuration."""
+        current_settings = self.cluster_config
+        # We use current_settings and new_conf and check for any differences
+        # therefore, we need to make a deepcopy here before editing new_conf
+        new_conf = copy.deepcopy(current_settings)
+        for key_to_del in config.config_entries_to_del:
+            if key_to_del in new_conf:
+                del new_conf[key_to_del]
+        new_conf |= config.config_entries_to_add
+
+        logger.debug(
+            "Difference between current and new configuration: \n"
+            + str(
+                {
+                    "in current but not in new_conf": {
+                        k: v for k, v in current_settings.items() if k not in new_conf.keys()
+                    },
+                    "in new_conf but not in current": {
+                        k: v for k, v in new_conf.items() if k not in current_settings.keys()
+                    },
+                    "in both but different values": {
+                        k: v
+                        for k, v in new_conf.items()
+                        if k in current_settings.keys() and current_settings[k] != v
+                    },
+                }
+            )
+        )
+        return current_settings, new_conf
+
     def apply_config(self, config: OpenSearchPluginConfig) -> bool:
         """Runs the configuration changes as passed via OpenSearchPluginConfig.
 
@@ -262,27 +309,25 @@ class OpenSearchPluginManager:
         1) Remove the entries to be deleted
         2) Add entries, if available
 
-        For example:
-            KNN needs to:
-            1) Remove from configuration: {"knn.plugin.enabled": "True"}
-            2) Add to configuration: {"knn.plugin.enabled": "False"}
-
         Returns True if a configuration change was performed.
         """
         self._keystore.delete(config.secret_entries_to_del)
         self._keystore.add(config.secret_entries_to_add)
-        # Add and remove configuration if applies
-        if config.config_entries_to_del:
-            self._opensearch_config.delete_plugin(config.config_entries_to_del)
-
-        if config.config_entries_to_add:
-            self._opensearch_config.add_plugin(config.config_entries_to_add)
-
         if config.secret_entries_to_del or config.secret_entries_to_add:
             self._keystore.reload_keystore()
 
-        # Return True if some configuration entries changed
-        return True if config.config_entries_to_add or config.config_entries_to_del else False
+        current_settings, new_conf = self._compute_settings(config)
+        if current_settings == new_conf:
+            # Nothing to do here
+            logger.info("apply_config: nothing to do, return")
+            return False
+
+        # Update the configuration
+        if config.config_entries_to_del:
+            self._opensearch_config.delete_plugin(config.config_entries_to_del)
+        if config.config_entries_to_add:
+            self._opensearch_config.add_plugin(config.config_entries_to_add)
+        return True
 
     def status(self, plugin: OpenSearchPlugin) -> PluginState:
         """Returns the status for a given plugin."""
@@ -320,32 +365,28 @@ class OpenSearchPluginManager:
     def _is_enabled(self, plugin: OpenSearchPlugin) -> bool:
         """Returns true if plugin is enabled.
 
-        The main question to answer is if we would remove the configuration from
-        opensearch.yaml and/or add some other configuration instead.
-        If yes, then we know that the service is enabled.
+        The main question to answer is if we would have it in the configuration
+        from cluster settings. If yes, then we know that the service is enabled.
 
-        The disable() method is used, as a given entry will be removed from opensearch.yml.
-
-        If disable() returns a list, then check if the keys in the stored_plugin_conf
-        matches the list values; otherwise, if a dict, then check if the keys and values match.
-        If they match in any of the cases above, then return True.
+        Check if the configuration from enable() is present or not.
         """
         try:
-            plugin_conf = plugin.disable().config_entries_to_del
-            keystore_conf = plugin.disable().secret_entries_to_del
-            stored_plugin_conf = self._opensearch_config.get_plugin(plugin_conf)
-
-            if any(k not in self._keystore.list() for k in keystore_conf):
+            current_settings, new_conf = self._compute_settings(plugin.config())
+            if current_settings != new_conf:
                 return False
 
-            # Using sets to guarantee matches; stored_plugin_conf will be a dict
-            if isinstance(plugin_conf, list):
-                return set(plugin_conf) == set(stored_plugin_conf)
-            else:  # plugin_conf is a dict
-                return plugin_conf == stored_plugin_conf
+            # Now, focus on the keystore part
+            keys_available = self._keystore.list()
+            keys_to_add = plugin.config().secret_entries_to_add
+            if any(k not in keys_available for k in keys_to_add):
+                return False
+            keys_to_del = plugin.config().secret_entries_to_del
+            if any(k in keys_available for k in keys_to_del):
+                return False
         except (KeyError, OpenSearchPluginError) as e:
             logger.warning(f"_is_enabled: error with {e}")
             return False
+        return True
 
     def _needs_upgrade(self, plugin: OpenSearchPlugin) -> bool:
         """Returns true if plugin needs upgrade."""

--- a/lib/charms/opensearch/v0/opensearch_plugins.py
+++ b/lib/charms/opensearch/v0/opensearch_plugins.py
@@ -277,6 +277,7 @@ from typing import Any, Dict, List, Optional
 from charms.opensearch.v0.helper_enums import BaseStrEnum
 from charms.opensearch.v0.opensearch_exceptions import OpenSearchError
 from jproperties import Properties
+from pydantic import BaseModel, validator
 
 # The unique Charmhub library identifier, never change it
 LIBID = "3b05456c6e304680b4af8e20dae246a2"
@@ -331,20 +332,28 @@ class PluginState(BaseStrEnum):
     WAITING_FOR_UPGRADE = "waiting-for-upgrade"
 
 
-class OpenSearchPluginConfig:
-    """Represent the configuration of a plugin to be applied when configuring or disabling it."""
+class OpenSearchPluginConfig(BaseModel):
+    """Represent the configuration of a plugin to be applied when configuring or disabling it.
 
-    def __init__(
-        self,
-        config_entries_to_add: Optional[Dict[str, str]] = None,
-        config_entries_to_del: Optional[List[str]] = None,
-        secret_entries_to_add: Optional[Dict[str, str]] = None,
-        secret_entries_to_del: Optional[List[str]] = None,
-    ):
-        self.config_entries_to_add = config_entries_to_add or {}
-        self.config_entries_to_del = config_entries_to_del or []
-        self.secret_entries_to_add = secret_entries_to_add or {}
-        self.secret_entries_to_del = secret_entries_to_del or []
+    The config may receive any type of data, but will convert everything to strings and
+    pay attention to special types, such as booleans, which need to be "true" or "false".
+    """
+
+    config_entries_to_add: Optional[Dict[str, str]] = {}
+    config_entries_to_del: Optional[List[str]] = []
+    secret_entries_to_add: Optional[Dict[str, str]] = {}
+    secret_entries_to_del: Optional[List[str]] = []
+
+    @validator("config_entries_to_add", "secret_entries_to_add", allow_reuse=True, pre=True)
+    def convert_values_to_add(cls, conf) -> Dict[str, str]:  # noqa N805
+        """Converts the object to a dictionary.
+
+        Respects the conversion for boolean to {"true", "false"}.
+        """
+        return {
+            key: str(val).lower() if isinstance(val, bool) else str(val)
+            for key, val in conf.items()
+        }
 
 
 class OpenSearchPlugin:
@@ -389,9 +398,9 @@ class OpenSearchPlugin:
         Format:
         OpenSearchPluginConfig(
             config_entries_to_add = {...},
-            config_entries_to_del = [...]|{...},
+            config_entries_to_del = [...],
             secret_entries_to_add = {...},
-            secret_entries_to_del = [...]|{...},
+            secret_entries_to_del = [...],
         )
 
         May throw KeyError if accessing some source, such as self._extra_config, but the
@@ -406,9 +415,9 @@ class OpenSearchPlugin:
         Format:
         OpenSearchPluginConfig(
             config_entries_to_add = {...},
-            config_entries_to_del = [...]|{...},
+            config_entries_to_del = [...],
             secret_entries_to_add = {...},
-            secret_entries_to_del = [...]|{...},
+            secret_entries_to_del = [...],
         )
 
         May throw KeyError if accessing some source, such as self._extra_config, but the
@@ -429,14 +438,12 @@ class OpenSearchKnn(OpenSearchPlugin):
         """Returns a plugin config object to be applied for enabling the current plugin."""
         return OpenSearchPluginConfig(
             config_entries_to_add={"knn.plugin.enabled": True},
-            config_entries_to_del={"knn.plugin.enabled": False},
         )
 
     def disable(self) -> OpenSearchPluginConfig:
         """Returns a plugin config object to be applied for disabling the current plugin."""
         return OpenSearchPluginConfig(
             config_entries_to_add={"knn.plugin.enabled": False},
-            config_entries_to_del={"knn.plugin.enabled": True},
         )
 
     @property
@@ -462,9 +469,9 @@ class OpenSearchBackupPlugin(OpenSearchPlugin):
         Format:
         OpenSearchPluginConfig(
             config_entries_to_add = {...},
-            config_entries_to_del = [...]|{...},
+            config_entries_to_del = [...],
             secret_entries_to_add = {...},
-            secret_entries_to_del = [...]|{...},
+            secret_entries_to_del = [...],
         )
         """
         if not self._extra_config.get("access-key") or not self._extra_config.get("secret-key"):
@@ -480,11 +487,6 @@ class OpenSearchBackupPlugin(OpenSearchPlugin):
             )
 
         return OpenSearchPluginConfig(
-            # Try to remove the previous values
-            secret_entries_to_del=[
-                "s3.client.default.access_key",
-                "s3.client.default.secret_key",
-            ],
             secret_entries_to_add={
                 # Remove any entries with None value
                 k: v
@@ -502,9 +504,9 @@ class OpenSearchBackupPlugin(OpenSearchPlugin):
         Format:
         OpenSearchPluginConfig(
             config_entries_to_add = {...},
-            config_entries_to_del = [...]|{...},
+            config_entries_to_del = [...],
             secret_entries_to_add = {...},
-            secret_entries_to_del = [...]|{...},
+            secret_entries_to_del = [...],
         )
         """
         return OpenSearchPluginConfig(

--- a/lib/charms/opensearch/v0/opensearch_relation_provider.py
+++ b/lib/charms/opensearch/v0/opensearch_relation_provider.py
@@ -328,18 +328,18 @@ class OpenSearchProvider(Object):
         """
         if not self.unit.is_leader():
             # We're updating app databags in this function, so it can't be called on follower
-            # units.
+            # units. This is not checked in `set_tls_ca`, in data-interfaces.
             return
-        if not ca_chain:
-            try:
-                ca_chain = self.charm.secrets.get_object(Scope.APP, CertType.APP_ADMIN.val).get(
-                    "chain"
-                )
-            except AttributeError:
-                # cert doesn't exist - presumably we don't yet have a TLS relation.
-                return
-
-        self.opensearch_provides.set_tls_ca(relation_id, ca_chain)
+        try:
+            # If the ca_chain=None, then we try to load the CA from the app-admin secret.
+            _ch_chain = ca_chain or self.charm.secrets.get_object(
+                Scope.APP, CertType.APP_ADMIN.val
+            ).get("chain")
+        except AttributeError:
+            # cert doesn't exist - presumably we don't yet have a TLS relation.
+            logger.warning("unable to get ca_chain")
+            return
+        self.opensearch_provides.set_tls_ca(relation_id, _ch_chain)
 
     def _on_relation_changed(self, event: RelationChangedEvent) -> None:
         if not self.unit.is_leader():

--- a/tests/integration/ha/test_large_deployments.py
+++ b/tests/integration/ha/test_large_deployments.py
@@ -120,6 +120,8 @@ async def test_set_roles_manually(
         assert node.temperature is None, "Node temperature was erroneously set."
 
     # change cluster name and roles + temperature, should trigger a rolling restart
+
+    logger.info("Changing cluster name and roles + temperature.")
     await ops_test.model.applications[app].set_config(
         {"cluster_name": "new_cluster_name", "roles": "cluster_manager, data.cold"}
     )
@@ -131,6 +133,8 @@ async def test_set_roles_manually(
         wait_for_exact_units=len(nodes),
         idle_period=IDLE_PERIOD,
     )
+
+    logger.info("Checking if the cluster name and roles + temperature were changed.")
     assert await check_cluster_formation_successful(
         ops_test, leader_unit_ip, get_application_unit_names(ops_test, app=app)
     )

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -79,6 +79,7 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
     assert len(ops_test.model.applications[APP_NAME].units) == 3
 
 
+@pytest.mark.abort_on_fail
 async def test_prometheus_exporter_enabled_by_default(ops_test):
     """Test that Prometheus Exporter is running before the relation is there."""
     leader_unit_ip = await get_leader_unit_ip(ops_test, app=APP_NAME)
@@ -90,6 +91,7 @@ async def test_prometheus_exporter_enabled_by_default(ops_test):
     assert len(response_str.split("\n")) > 500
 
 
+@pytest.mark.abort_on_fail
 async def test_prometheus_exporter_cos_relation(ops_test):
     await ops_test.model.deploy(COS_APP_NAME, channel="edge"),
     await ops_test.model.integrate(APP_NAME, COS_APP_NAME)
@@ -119,6 +121,7 @@ async def test_prometheus_exporter_cos_relation(ops_test):
     assert relation_data["scheme"] == "https"
 
 
+@pytest.mark.abort_on_fail
 async def test_monitoring_user_fetch_prometheus_data(ops_test):
     leader_unit_ip = await get_leader_unit_ip(ops_test, app=APP_NAME)
     endpoint = f"https://{leader_unit_ip}:9200/_prometheus/metrics"
@@ -139,6 +142,7 @@ async def test_monitoring_user_fetch_prometheus_data(ops_test):
     assert len(response_str.split("\n")) > 500
 
 
+@pytest.mark.abort_on_fail
 async def test_prometheus_monitor_user_password_change(ops_test):
     # Password change applied as expected
     leader_id = await get_leader_unit_id(ops_test, APP_NAME)
@@ -161,6 +165,7 @@ async def test_prometheus_monitor_user_password_change(ops_test):
     assert relation_data["password"] == new_password
 
 
+@pytest.mark.abort_on_fail
 async def test_knn_enabled_disabled(ops_test):
     config = await ops_test.model.applications[APP_NAME].get_config()
     assert config["plugin_opensearch_knn"]["default"] is True
@@ -177,6 +182,7 @@ async def test_knn_enabled_disabled(ops_test):
 
     config = await ops_test.model.applications[APP_NAME].get_config()
     assert config["plugin_opensearch_knn"]["value"] is True
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", idle_period=45)
 
 
 @pytest.mark.abort_on_fail
@@ -218,7 +224,7 @@ async def test_knn_search_with_hnsw_faiss(ops_test: OpsTest) -> None:
     # Insert data in bulk
     await bulk_insert(ops_test, app, leader_unit_ip, payload)
     query = {"size": 2, "query": {"knn": {vector_name: {"vector": payload_list[0], "k": 2}}}}
-    docs = await search(ops_test, app, leader_unit_ip, index_name, query)
+    docs = await search(ops_test, app, leader_unit_ip, index_name, query, retries=30)
     assert len(docs) == 2
 
 
@@ -261,7 +267,7 @@ async def test_knn_search_with_hnsw_nmslib(ops_test: OpsTest) -> None:
     # Insert data in bulk
     await bulk_insert(ops_test, app, leader_unit_ip, payload)
     query = {"size": 2, "query": {"knn": {vector_name: {"vector": payload_list[0], "k": 2}}}}
-    docs = await search(ops_test, app, leader_unit_ip, index_name, query)
+    docs = await search(ops_test, app, leader_unit_ip, index_name, query, retries=30)
     assert len(docs) == 2
 
 

--- a/tests/unit/lib/test_backups.py
+++ b/tests/unit/lib/test_backups.py
@@ -490,12 +490,18 @@ class TestBackups(unittest.TestCase):
         assert self.charm.backup._get_endpoint_protocol("https://10.0.0.2:8000") == "https"
         assert self.charm.backup._get_endpoint_protocol("test.not-valid-url") == "https"
 
+    @patch(
+        "charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager.check_plugin_manager_ready"
+    )
     @patch("charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager.status")
     @patch("charms.opensearch.v0.opensearch_backups.OpenSearchBackup.apply_api_config_if_needed")
     @patch("charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager.apply_config")
     @patch("charms.opensearch.v0.opensearch_distro.OpenSearchDistribution.version")
-    def test_00_update_relation_data(self, __, mock_apply_config, _, mock_status) -> None:
+    def test_00_update_relation_data(
+        self, __, mock_apply_config, _, mock_status, mock_pm_ready
+    ) -> None:
         """Tests if new relation without data returns."""
+        mock_pm_ready.return_value = True
         mock_status.return_value = PluginState.INSTALLED
         self.harness.update_relation_data(
             self.s3_rel_id,
@@ -513,10 +519,6 @@ class TestBackups(unittest.TestCase):
         assert (
             mock_apply_config.call_args[0][0].__dict__
             == OpenSearchPluginConfig(
-                secret_entries_to_del=[
-                    "s3.client.default.access_key",
-                    "s3.client.default.secret_key",
-                ],
                 secret_entries_to_add={
                     "s3.client.default.access_key": "aaaa",
                     "s3.client.default.secret_key": "bbbb",

--- a/tests/unit/lib/test_helper_cluster.py
+++ b/tests/unit/lib/test_helper_cluster.py
@@ -241,3 +241,38 @@ class TestHelperCluster(unittest.TestCase):
         self.assertEqual(raw_node.name, from_json_node.name)
         self.assertEqual(raw_node.roles, from_json_node.roles)
         self.assertEqual(raw_node.ip, from_json_node.ip)
+
+    @patch("charms.opensearch.v0.helper_cluster.OpenSearchDistribution.request")
+    def test_get_cluster_settings(self, request_mock):
+        """Test the get_cluster_settings method."""
+        request_mock.return_value = {
+            "defaults": {
+                "knn.plugin.enabled": "false",
+            },
+            "persistent": {
+                "knn.plugin.enabled": "true",
+                "cluster.routing.allocation.enable": "all",
+            },
+            "transient": {
+                "indices.recovery.max_bytes_per_sec": "50mb",
+            },
+        }
+
+        expected_settings = {
+            "knn.plugin.enabled": "true",
+            "cluster.routing.allocation.enable": "all",
+            "indices.recovery.max_bytes_per_sec": "50mb",
+        }
+
+        settings = ClusterTopology.get_cluster_settings(
+            self.opensearch,
+            include_defaults=True,
+        )
+
+        self.assertEqual(settings, expected_settings)
+        request_mock.assert_called_once_with(
+            "GET",
+            "/_cluster/settings?flat_settings=true&include_defaults=true",
+            host=None,
+            alt_hosts=None,
+        )

--- a/tests/unit/lib/test_ml_plugins.py
+++ b/tests/unit/lib/test_ml_plugins.py
@@ -54,7 +54,12 @@ class TestOpenSearchKNN(unittest.TestCase):
         }
         self.charm.opensearch.is_started = MagicMock(return_value=True)
         self.charm.health.apply = MagicMock(return_value=HealthColors.GREEN)
+        self.plugin_manager._is_cluster_ready = MagicMock(return_value=True)
+        charms.opensearch.v0.helper_cluster.ClusterTopology.get_cluster_settings = MagicMock(
+            return_value={}
+        )
 
+    @patch(f"{BASE_LIB_PATH}.opensearch_distro.OpenSearchDistribution.is_node_up")
     @patch(
         f"{BASE_LIB_PATH}.opensearch_peer_clusters.OpenSearchPeerClustersManager.deployment_desc"
     )
@@ -78,6 +83,7 @@ class TestOpenSearchKNN(unittest.TestCase):
         mock_version,
         mock_acquire_lock,
         ___,
+        mock_is_node_up,
     ) -> None:
         """Tests entire config_changed event with KNN plugin."""
         mock_status.return_value = PluginState.ENABLED
@@ -89,12 +95,12 @@ class TestOpenSearchKNN(unittest.TestCase):
         self.plugin_manager._opensearch_config.delete_plugin = MagicMock()
         self.plugin_manager._opensearch_config.add_plugin = MagicMock()
         self.charm.status = MagicMock()
+        mock_is_node_up.return_value = True
+        self.charm._get_nodes = MagicMock(return_value=[1])
+        self.charm.planned_units = MagicMock(return_value=1)
 
         self.harness.update_config({"plugin_opensearch_knn": False})
         mock_acquire_lock.assert_called_once()
-        self.plugin_manager._opensearch_config.delete_plugin.assert_called_once_with(
-            {"knn.plugin.enabled": True}
-        )
         self.plugin_manager._opensearch_config.add_plugin.assert_called_once_with(
-            {"knn.plugin.enabled": False}
+            {"knn.plugin.enabled": "false"}
         )

--- a/tests/unit/lib/test_plugins.py
+++ b/tests/unit/lib/test_plugins.py
@@ -122,6 +122,7 @@ class TestOpenSearchPlugin(unittest.TestCase):
         self.charm.opensearch.is_started = MagicMock(return_value=True)
         self.charm.health.apply = MagicMock(return_value=HealthColors.GREEN)
         self.charm.opensearch.version = "2.9.0"
+        self.plugin_manager._is_cluster_ready = MagicMock(return_value=True)
 
     @patch("charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager._is_enabled")
     @patch("charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager._is_installed")
@@ -242,6 +243,16 @@ class TestOpenSearchPlugin(unittest.TestCase):
         # Mock _installed_plugins to return test
         mock_installed_plugins.return_value = ["test"]
 
+        self.charm._get_nodes = MagicMock(
+            return_value={
+                "1": {},
+                "2": {},
+                "3": {},
+            }
+        )
+        self.charm.app.planned_units = MagicMock(return_value=3)
+        self.charm.opensearch.is_node_up = MagicMock(return_value=True)
+
         mock_load.return_value = {}
         # run is called, but only _configure method really matter:
         # Set install to false, so only _configure is evaluated
@@ -291,6 +302,16 @@ class TestOpenSearchPlugin(unittest.TestCase):
         # Return a fake content of the relation
         mock_process_relation.return_value = {"param": "tested"}
 
+        self.charm._get_nodes = MagicMock(
+            return_value={
+                "1": {},
+                "2": {},
+                "3": {},
+            }
+        )
+        self.charm.app.planned_units = MagicMock(return_value=3)
+        self.charm.opensearch.is_node_up = MagicMock(return_value=True)
+
         # Keystore-related mocks
         self.plugin_manager._keystore._add = MagicMock()
         self.plugin_manager._opensearch.request = MagicMock(return_value={"status": 200})
@@ -311,11 +332,13 @@ class TestOpenSearchPlugin(unittest.TestCase):
         mock_plugin_relation.return_value = True
         # plugin is initially disabled and enabled when method self._disable calls self.status
         mock_is_enabled.side_effect = [
+            False,  # called by logger
             False,  # called by self.status, in self._install
             False,  # called by self._configure
             True,  # called by self.status, in self._disable
+            True,  # called by logger
         ]
-
+        charms.opensearch.v0.opensearch_plugin_manager.logger = MagicMock()
         self.assertTrue(self.plugin_manager.run())
         self.plugin_manager._keystore._add.assert_has_calls([call("key1", "secret1")])
         self.charm.opensearch.config.put.assert_has_calls(
@@ -326,6 +349,7 @@ class TestOpenSearchPlugin(unittest.TestCase):
             [call("POST", "_nodes/reload_secure_settings")]
         )
 
+    @patch("charms.opensearch.v0.opensearch_plugin_manager.ClusterTopology.get_cluster_settings")
     @patch("charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager._extra_conf")
     @patch("charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager._is_enabled")
     @patch(
@@ -334,16 +358,15 @@ class TestOpenSearchPlugin(unittest.TestCase):
     @patch(
         "charms.opensearch.v0.opensearch_plugin_manager.OpenSearchPluginManager._installed_plugins"
     )
-    @patch("charms.opensearch.v0.opensearch_config.OpenSearchConfig.load_node")
     @patch("charms.opensearch.v0.opensearch_distro.OpenSearchDistribution.version")
     def test_disable_plugin(
         self,
         _,
-        mock_load,
         mock_installed_plugins,
         mock_plugin_relation,
         mock_is_enabled,
-        mock_extra_conf,
+        __,
+        mock_get_cluster_settings,
     ) -> None:
         """Tests end-to-end the disable of a plugin."""
         # Keystore-related mocks
@@ -363,8 +386,17 @@ class TestOpenSearchPlugin(unittest.TestCase):
         # Mock _installed_plugins to return test
         mock_installed_plugins.return_value = ["test"]
 
-        # load_node will be called multiple times
-        mock_load.side_effect = {"param": "tested"}
+        self.charm._get_nodes = MagicMock(
+            return_value={
+                "1": {},
+                "2": {},
+                "3": {},
+            }
+        )
+        self.charm.app.planned_units = MagicMock(return_value=3)
+        self.charm.opensearch.is_node_up = MagicMock(return_value=True)
+
+        mock_get_cluster_settings.return_value = {"param": "tested"}
         mock_plugin_relation.return_value = False
         # plugin is initially disabled and enabled when method self._disable calls self.status
         mock_is_enabled.return_value = True
@@ -414,10 +446,6 @@ class TestOpenSearchBackupPlugin(unittest.TestCase):
             "secret-key": "SECRET_KEY",
         }
         expected_config = OpenSearchPluginConfig(
-            secret_entries_to_del=[
-                "s3.client.default.access_key",
-                "s3.client.default.secret_key",
-            ],
             secret_entries_to_add={
                 "s3.client.default.access_key": "ACCESS_KEY",
                 "s3.client.default.secret_key": "SECRET_KEY",

